### PR TITLE
Use new nvCOMP API to optimize the decompression temp memory size

### PR DESCRIPTION
**Pull request by vuule**
nvCOMP 2.3.1 introduces expanded API to determine the amount of temporary memory required to decompress a set of chunks. New API also takes the maximum total size of the decompressed data for all chunks.
This PR expands the nvCOMP adapter to use this API when available (nvCOMP version >= 2.3.1). With older versions the older API is still used and there is no behavior change.
ORC and Parquet reader now pass the total decompress data size to the nvCOMP adapter decompression calls.

### CODE REVIEW

**Comment by vuule** [comment]
> Not a part of this PR but I see that a change was introduced in #10699 that reverted the changes made in #9995. The latter added a custom kernel to check statuses to avoid using `thrust::all_of` which supposedly caused a device synchronize and deteriorated perf when running multiple parquet read calls on the same device.

What do you think about addressing this in a separate PR? Should help merge it faster.

File: cpp/src/io/comp/nvcomp_adapter.cpp [hunk 1]
```diff
@@ -37,8 +37,36 @@
...
+      return nvcompBatchedSnappyDecompressGetTempSizeEx(std::forward<Args>(args)...);
+#if NVCOMP_HAS_ZSTD
+    case compression_type::ZSTD:
+      return nvcompBatchedZstdDecompressGetTempSizeEx(std::forward<Args>(args)...);
+#endif
+#if NVCOMP_HAS_DEFLATE
+    case compression_type::DEFLATE:
+      return nvcompBatchedDeflateDecompressGetTempSizeEx(std::forward<Args>(args)...);
+#endif
+    default: CUDF_FAIL("Unsupported compression type");
```
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/comp/nvcomp_adapter.cpp:65]: the compression type enum in the nvcomp adapter does not have NONE. We can add it if/when we have a higher level API that encapsulates both nvCOMP and internal implementations.

File: cpp/src/io/comp/uncomp.cpp [hunk 2]
```diff
@@ -522,8 +522,13 @@ size_t decompress_zstd(host_span<uint8_t const> src,
 
   auto hd_stats                   = hostdevice_vector<decompress_status>(1, stream);
   auto const max_uncomp_page_size = dst.size();
-  nvcomp::batched_decompress(
-    nvcomp::compression_type::ZSTD, hd_srcs, hd_dsts, hd_stats, max_uncomp_page_size, stream);
+  nvcomp::batched_decompress(nvcomp::compression_type::ZSTD,
```
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/comp/uncomp.cpp:525]: Agreed, nvCOMP is highly unlikely to be faster here. 
I used nvCOMP to expedite the development and avoid adding another dependency. I saw that the footer size is much smaller than the file size, so the relative perf impact should be minimal.
Opened issue https://github.com/rapidsai/cudf/issues/11071 to investigate and replace with a CPU library.

**Review by devavret — APPROVED** [review]

**Review by jbrennan333 — APPROVED** [review]
+1 this looks good to me.  I have built it and tested with local nvcomp-2.3.1 libs.

File: python/cudf/cudf/_lib/cpp/io/types.pxd [hunk 3]
```diff
@@ -32,6 +32,10 @@ cdef extern from "cudf/io/types.hpp" \
         BROTLI "cudf::io::compression_type::BROTLI"
         ZIP "cudf::io::compression_type::ZIP"
         XZ "cudf::io::compression_type::XZ"
+        ZLIB "cudf::io::compression_type::ZLIB"
+        LZ4 "cudf::io::compression_type::LZ4"
+        LZO "cudf::io::compression_type::LZO"
+        ZSTD "cudf::io::compression_type::ZSTD"
```
Comment (vyasr) [inline python/cudf/cudf/_lib/cpp/io/types.pxd:38]: There's significant divergence between the set of compressions supported at the C++ level (as defined by this enum) and the ones exported to Python (in orc.pyx). Is that based on the what compression types are supported by a given file format (e.g. ORC) or are we just choosing not to expose all features?

**Review by vyasr — APPROVED** [review]
Python approval (didn't review the CPP).

File: python/cudf/cudf/_lib/cpp/io/types.pxd (hunk 3 above)
Comment (vuule) [reply python/cudf/cudf/_lib/cpp/io/types.pxd:38]: Yeah, this covers all compression types in all of cuIO. Different formats only support a subset (and we support a subset of that).

**Review by bdice — APPROVED** [review]
All looks good to me!

**Pull request merged**